### ISSUE TITLE
[PATCH] Locale::Maketext - update use of I18N::LangTags to happen at load time.

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10507]
Migrated from [rt.perl.org#76668](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=76668) (status was 'resolved')

Searchable as RT76668$

**Comment by p5pRT** [comment]
### From @toddr

#### Created by @toddr



This patch with tests resolves CPAN RT \#48808\. 

This change&#8203;:
1\. loads I18N&#8203;::LangTags&#8203;::Detect in BEGIN rather than whenever \_ambient\_langprefs is called
2\. Removes the version number requirement for I18N&#8203;::LangTags\. Detect didn't come around until 0\.31 so a dep on 0\.30 is unnecessary\. 

This will assure any load issues with I18N&#8203;::LangTags&#8203;::Detect happen at BEGIN\, not randomly somewhere in the program\. 

This does increase the dependency on I18N&#8203;::LangTags from 0\.30 to 0\.31\, but the release dates are as follows\, so I don't think it's a big deal&#8203;:

0\.30 \- 2004\-03\-30  Sean M\. Burke  sburke@&#8203;cpan\.org
0\.31 \- 2004\-06\-17  Sean M\. Burke  sburke@&#8203;cpan\.org

<details><summary>Perl Info</summary>

```

Flags:
   category=library
   severity=low
   module=Locale::Maketext
   Type=Patch
   PatchStatus=HasPatch

Site configuration information for perl 5.12.1:

Configured by cPanel at Mon Jun  7 11:16:28 CDT 2010.

Summary of my perl5 (revision 5 version 12 subversion 1) configuration:

 Platform:
   osname=linux, osvers=2.6.18-194.3.1.el5, archname=x86_64-linux
   uname='linux rpmb-centos-50-64bit 2.6.18-194.3.1.el5 #1 smp thu may 13 13:08:30 edt 2010 x86_64 x86_64 x86_64 gnulinux '
   config_args='-des -Darchname=x86_64-linux -Dcc=/usr/local/cpanel/bin/gcc -Dcpp=/usr/local/cpanel/bin/gcc -E -DDEBUGGING=none -Doptimize=-Os -Dusemymalloc=y -Duseshrplib=true -Duselargefiles=yes -Duseposix=true -Dhint=recommended -Duseperlio=yes -Dccflags=-I/usr/local/cpanel/include -L/usr/local/cpanel/lib64 -Wl,-rpath -Wl,/usr/local/cpanel/lib64  -Dcppflags=-I/usr/local/cpanel/include -L/usr/local/cpanel/lib64 -Dldflags=-Wl,-rpath -Wl,/usr/local/cpanel/lib64 -L/usr/local/cpanel/lib64 -Dprefix=/usr/local/cpanel -Dsiteprefix=/usr/local/cpanel -Dsitebin=/usr/local/cpanel/bin -Dsitelib=/usr/local/cpanel/lib64/perl5/site_lib -Dprivlib=/usr/local/cpanel/lib64/perl5/5.12.1 -Dotherlibdirs=/var/cpanel/perl5/lib:/usr/local/cpanel/lib64/perl5/cpanel -Dman1dir=/usr/local/cpanel/share/man/man1 -Dman3dir=/usr/local/cpanel/share/man/man3 -Dsiteman1dir=/usr/local/cpanel/share/man/man1 -Dsiteman3dir=/usr/local/cpanel/share/man/man3 -Dcf_by=cPanel -Dmyhostname=localhost -Dperladmin=root@localhost -Dcf_email=support@cpanel.net -Di_dbm=/usr/local/cpanel/include -Di_gdbm=/usr/local/cpanel/include -Di_ndbm=/usr/local/cpanel/include -Ud_dosuid -Uuserelocatableinc -Umad -Uusethreads -Uusemultiplicity -Uusesocks -Uuselongdouble -Ui_db -Aldflags=-L/usr/local/cpanel/lib64 -L/usr/lib64 -L/lib64 -lgdbm -Dlocincpth=/usr/local/cpanel/include /usr/local/include  -Acflags=-fPIC -DPIC -m64 -I/usr/local/cpanel/include -Dlibpth=/usr/local/cpanel/lib64 /usr/local/lib64 /usr/local/lib /lib64 /usr/lib64  -Duse64bitint=yes -Duse64bitall=yes'
   hint=recommended, useposix=true, d_sigaction=define
   useithreads=undef, usemultiplicity=undef
   useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
   use64bitint=define, use64bitall=define, uselongdouble=undef
   usemymalloc=y, bincompat5005=undef
 Compiler:
   cc='/usr/local/cpanel/bin/gcc', ccflags ='-I/usr/local/cpanel/include -L/usr/local/cpanel/lib64 -Wl,-rpath -Wl,/usr/local/cpanel/lib64 -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/cpanel/include -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
   optimize='-Os',
   cppflags='-I/usr/local/cpanel/include -L/usr/local/cpanel/lib64 -I/usr/local/cpanel/include -L/usr/local/cpanel/lib64 -Wl,-rpath -Wl,/usr/local/cpanel/lib64 -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/cpanel/include -I/usr/local/include'
   ccversion='', gccversion='4.5.0', gccosandvers=''
   intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
   d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
   ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
   alignbytes=8, prototype=define
 Linker and Libraries:
   ld='/usr/local/cpanel/bin/gcc', ldflags ='-Wl,-rpath -Wl,/usr/local/cpanel/lib64 -L/usr/local/cpanel/lib64 -L/usr/local/cpanel/lib64 -L/usr/lib64 -L/lib64 -lgdbm -fstack-protector -L/usr/local/lib'
   libpth=/usr/local/cpanel/lib64 /usr/local/lib64 /usr/local/lib /lib64 /usr/lib64
   libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
   perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
   libc=/lib/libc-2.5.so, so=so, useshrplib=true, libperl=libperl.so
   gnulibc_version='2.5'
 Dynamic Linking:
   dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/usr/local/cpanel/lib64/perl5/5.12.1/x86_64-linux/CORE'
   cccdlflags='-fPIC', lddlflags='-shared -Os -L/usr/local/cpanel/lib64 -L/usr/lib64 -L/lib64 -L/usr/local/lib -fstack-protector'

Locally applied patches:
   cPanel Patches


@INC for perl 5.12.1:
   /usr/local/cpanel/lib64/perl5/site_lib/x86_64-linux
   /usr/local/cpanel/lib64/perl5/site_lib
   /usr/local/cpanel/lib64/perl5/5.12.1/x86_64-linux
   /usr/local/cpanel/lib64/perl5/5.12.1
   /var/cpanel/perl5/lib
   /usr/local/cpanel/lib64/perl5/cpanel
   .


Environment for perl 5.12.1:
   HOME=/home/toddr
   LANG=en_US.UTF-8
   LANGUAGE (unset)
   LD_LIBRARY_PATH (unset)
   LOGDIR (unset)
   PATH=/usr/local/cpanel/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/usr/X11/bin:
   PERL_BADLANG (unset)
   SHELL=/bin/zsh


```
</details>

**Comment by p5pRT** [comment]
### From @toddr

<details><summary>Inline Patch</summary>

```diff
diff --git a/dist/Locale-Maketext/lib/Locale/Maketext.pm b/dist/Locale-Maketext/lib/Locale/Maketext.pm
index 929a70e..f4959fe 100644
--- a/dist/Locale-Maketext/lib/Locale/Maketext.pm
+++ b/dist/Locale-Maketext/lib/Locale/Maketext.pm
@@ -3,7 +3,8 @@ use strict;
 use vars qw( @ISA $VERSION $MATCH_SUPERS $USING_LANGUAGE_TAGS
 $USE_LITERALS $MATCH_SUPERS_TIGHTLY);
 use Carp ();
-use I18N::LangTags 0.30 ();
+use I18N::LangTags ();
+use I18N::LangTags::Detect ();
 
 #--------------------------------------------------------------------------
 
@@ -379,7 +380,6 @@ sub _langtag_munging {
 ###########################################################################
 
 sub _ambient_langprefs {
-    require I18N::LangTags::Detect;
     return  I18N::LangTags::Detect::detect();
 }
 
```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Jul 22 10&#8203;:33&#8203;:05 2010\, toddr@&#8203;cpanel\.net wrote&#8203;:
> This is a bug report for perl from toddr@&#8203;cpanel\.net\,
> generated with the help of perlbug 1\.39 running under perl 5\.12\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> This patch with tests resolves CPAN RT \#48808\.

What tests? :\-\)

> 
> This change&#8203;:
> 1\. loads I18N&#8203;::LangTags&#8203;::Detect in BEGIN rather than whenever
>    \_ambient\_langprefs is called
> 2\. Removes the version number requirement for I18N&#8203;::LangTags\. Detect
>    didn't come around until 0\.31 so a dep on 0\.30 is unnecessary\.
> 
> This will assure any load issues with I18N&#8203;::LangTags&#8203;::Detect happen at
>    BEGIN\, not randomly somewhere in the program\.
> 
> This does increase the dependency on I18N&#8203;::LangTags from 0\.30 to 0\.31\,
>    but the release dates are as follows\, so I don't think it's a big
>    deal&#8203;:
> 
> 0\.30 \- 2004\-03\-30  Sean M\. Burke  sburke@&#8203;cpan\.org
> 0\.31 \- 2004\-06\-17  Sean M\. Burke  sburke@&#8203;cpan\.org
> 
> 
> 
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>    category=library
>    severity=low
>    module=Locale&#8203;::Maketext
>    Type=Patch
>    PatchStatus=HasPatch
> \-\-\-
> Site configuration information for perl 5\.12\.1&#8203;:
> 
> Configured by cPanel at Mon Jun  7 11&#8203;:16&#8203;:28 CDT 2010\.
> 
> Summary of my perl5 \(revision 5 version 12 subversion 1\)
>    configuration&#8203;:
> 
>  Platform&#8203;:
>    osname=linux\, osvers=2\.6\.18\-194\.3\.1\.el5\, archname=x86\_64\-linux
>    uname='linux rpmb\-centos\-50\-64bit 2\.6\.18\-194\.3\.1\.el5 \#1 smp thu may
>    13 13&#8203;:08&#8203;:30 edt 2010 x86\_64 x86\_64 x86\_64 gnulinux '
>    config\_args='\-des \-Darchname=x86\_64\-linux
>    \-Dcc=/usr/local/cpanel/bin/gcc \-Dcpp=/usr/local/cpanel/bin/gcc \-E
>    \-DDEBUGGING=none \-Doptimize=\-Os \-Dusemymalloc=y \-Duseshrplib=true
>    \-Duselargefiles=yes \-Duseposix=true \-Dhint=recommended
>    \-Duseperlio=yes \-Dccflags=\-I/usr/local/cpanel/include
>    \-L/usr/local/cpanel/lib64 \-Wl\,\-rpath \-Wl\,/usr/local/cpanel/lib64
>    \-Dcppflags=\-I/usr/local/cpanel/include \-L/usr/local/cpanel/lib64
>    \-Dldflags=\-Wl\,\-rpath \-Wl\,/usr/local/cpanel/lib64
>    \-L/usr/local/cpanel/lib64 \-Dprefix=/usr/local/cpanel
>    \-Dsiteprefix=/usr/local/cpanel \-Dsitebin=/usr/local/cpanel/bin
>    \-Dsitelib=/usr/local/cpanel/lib64/perl5/site\_lib
>    \-Dprivlib=/usr/local/cpanel/lib64/perl5/5\.12\.1
>   
\-Dotherlibdirs=/var/cpanel/perl5/lib&#8203;:/usr/local/cpanel/lib64/perl5/cpanel
>    \-Dman1dir=/usr/local/cpanel/share/man/man1
>    \-Dman3dir=/usr/local/cpanel/share/man/man3
>    \-Dsiteman1dir=/usr/local/cpanel/share/man/man1
>    \-Dsiteman3dir=/usr/local/cpanel/share/man/man3 \-Dcf\_by=cPanel
>    \-Dmyhostname=localhost \-Dperladmin=root@&#8203;localhost
>    \-Dcf\_email=support@&#8203;cpanel\.net \-Di\_dbm=/usr/local/cpanel/include
>    \-Di\_gdbm=/usr/local/cpanel/include
>    \-Di\_ndbm=/usr/local/cpanel/include \-Ud\_dosuid \-Uuserelocatableinc
>    \-Umad \-Uusethreads \-Uusemultiplicity \-Uusesocks \-Uuselongdouble
>    \-Ui\_db \-Aldflags=\-L/usr/local/cpanel/lib64 \-L/usr/lib64 \-L/lib64
>    \-lgdbm \-Dlocincpth=/usr/local/cpanel/include /usr/local/include
>    \-Acflags=\-fPIC \-DPIC \-m64 \-I/usr/local/cpanel/include
>    \-Dlibpth=/usr/local/cpanel/lib64 /usr/local/lib64 /usr/local/lib
>    /lib64 /usr/lib64  \-Duse64bitint=yes \-Duse64bitall=yes'
>    hint=recommended\, useposix=true\, d\_sigaction=define
>    useithreads=undef\, usemultiplicity=undef
>    useperlio=define\, d\_sfio=undef\, uselargefiles=define\,
>    usesocks=undef
>    use64bitint=define\, use64bitall=define\, uselongdouble=undef
>    usemymalloc=y\, bincompat5005=undef
>  Compiler&#8203;:
>    cc='/usr/local/cpanel/bin/gcc'\, ccflags ='\-
>    I/usr/local/cpanel/include \-L/usr/local/cpanel/lib64 \-Wl\,\-rpath
>    \-Wl\,/usr/local/cpanel/lib64 \-fno\-strict\-aliasing \-pipe
>    \-fstack\-protector \-I/usr/local/cpanel/include \-I/usr/local/include
>    \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
>    optimize='\-Os'\,
>    cppflags='\-I/usr/local/cpanel/include \-L/usr/local/cpanel/lib64
>    \-I/usr/local/cpanel/include \-L/usr/local/cpanel/lib64 \-Wl\,\-rpath
>    \-Wl\,/usr/local/cpanel/lib64 \-fno\-strict\-aliasing \-pipe
>    \-fstack\-protector \-I/usr/local/cpanel/include \-I/usr/local/include'
>    ccversion=''\, gccversion='4\.5\.0'\, gccosandvers=''
>    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
>    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
>    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
>    lseeksize=8
>    alignbytes=8\, prototype=define
>  Linker and Libraries&#8203;:
>    ld='/usr/local/cpanel/bin/gcc'\, ldflags ='\-Wl\,\-rpath
>    \-Wl\,/usr/local/cpanel/lib64 \-L/usr/local/cpanel/lib64
>    \-L/usr/local/cpanel/lib64 \-L/usr/lib64 \-L/lib64 \-lgdbm
>    \-fstack\-protector \-L/usr/local/lib'
>    libpth=/usr/local/cpanel/lib64 /usr/local/lib64 /usr/local/lib
>    /lib64 /usr/lib64
>    libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc \-lgdbm\_compat
>    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
>    libc=/lib/libc\-2\.5\.so\, so=so\, useshrplib=true\, libperl=libperl\.so
>    gnulibc\_version='2\.5'
>  Dynamic Linking&#8203;:
>    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E
>    \-Wl\,\-rpath\,/usr/local/cpanel/lib64/perl5/5\.12\.1/x86\_64\-linux/CORE'
>    cccdlflags='\-fPIC'\, lddlflags='\-shared \-Os
>    \-L/usr/local/cpanel/lib64 \-L/usr/lib64 \-L/lib64 \-L/usr/local/lib
>    \-fstack\-protector'
> 
> Locally applied patches&#8203;:
>    cPanel Patches
> 
> \-\-\-
> @&#8203;INC for perl 5\.12\.1&#8203;:
>    /usr/local/cpanel/lib64/perl5/site\_lib/x86\_64\-linux
>    /usr/local/cpanel/lib64/perl5/site\_lib
>    /usr/local/cpanel/lib64/perl5/5\.12\.1/x86\_64\-linux
>    /usr/local/cpanel/lib64/perl5/5\.12\.1
>    /var/cpanel/perl5/lib
>    /usr/local/cpanel/lib64/perl5/cpanel
>    \.
> 
> \-\-\-
> Environment for perl 5\.12\.1&#8203;:
>    HOME=/home/toddr
>    LANG=en\_US\.UTF\-8
>    LANGUAGE \(unset\)
>    LD\_LIBRARY\_PATH \(unset\)
>    LOGDIR \(unset\)
>   
PATH=/usr/local/cpanel/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/bin&#8203;:/usr/X11/bin&#8203;:
>    PERL\_BADLANG \(unset\)
>    SHELL=/bin/zsh
>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @toddr


On Sep 27\, 2010\, at 12&#8203;:06 AM\, Father Chrysostomos via RT wrote&#8203;:

> On Thu Jul 22 10&#8203;:33&#8203;:05 2010\, toddr@&#8203;cpanel\.net wrote&#8203;:
>> This is a bug report for perl from toddr@&#8203;cpanel\.net\,
>> generated with the help of perlbug 1\.39 running under perl 5\.12\.1\.
>> 
>> 
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \[Please describe your issue here\]
>> 
>> This patch with tests resolves CPAN RT \#48808\.
> 
> What tests? :\-\)

My bad\. No tests\. Do you think it's needed for this change?

Also I'm adding a git patch to make your life easier to commit it\.

**Comment by p5pRT** [comment]
### From @toddr

commit e0eafb09801c7ebdbf5f7769d4f00821dee08f90
Author&#8203;: Todd Rinaldo \<toddr@&#8203;cpan\.org>
Date&#8203;:   Tue Sep 28 12&#8203;:19&#8203;:58 2010 \-0500

&nbsp;    perl RT\# 76668 \- This patch resolves CPAN RT \#48808\.
&nbsp;    This change&#8203;:
&nbsp;    1\. loads I18N&#8203;::LangTags&#8203;::Detect in BEGIN rather than whenever \_ambient\_langprefs is called
&nbsp;    2\. Removes the version number requirement for I18N&#8203;::LangTags\. Detect didn't come around until 0\.31 so a dep on 0\.30 is unnecessary\.
&nbsp;    
&nbsp;    This will assure any load issues with I18N&#8203;::LangTags&#8203;::Detect happen at BEGIN\, not randomly somewhere in the program\.
&nbsp;    
&nbsp;    This does increase the dependency on I18N&#8203;::LangTags from 0\.30 to 0\.31\, but the release dates are as follows\, so I don't think it's a big deal&#8203;:
&nbsp;    
&nbsp;    0\.30 \- 2004\-03\-30 Sean M\. Burke sburke@&#8203;cpan\.org
&nbsp;    0\.31 \- 2004\-06\-17 Sean M\. Burke sburke@&#8203;cpan\.org

<details><summary>Inline Patch</summary>

```diff
diff --git a/dist/Locale-Maketext/lib/Locale/Maketext.pm b/dist/Locale-Maketext/lib/Locale/Maketext.pm
index cbfcb14..0f47d36 100644
--- a/dist/Locale-Maketext/lib/Locale/Maketext.pm
+++ b/dist/Locale-Maketext/lib/Locale/Maketext.pm
@@ -4,7 +4,8 @@ use strict;
 use vars qw( @ISA $VERSION $MATCH_SUPERS $USING_LANGUAGE_TAGS
 $USE_LITERALS $MATCH_SUPERS_TIGHTLY);
 use Carp ();
-use I18N::LangTags 0.30 ();
+use I18N::LangTags ();
+use I18N::LangTags::Detect ();
 
 #--------------------------------------------------------------------------
 
@@ -380,7 +381,6 @@ sub _langtag_munging {
 ###########################################################################
 
 sub _ambient_langprefs {
-    require I18N::LangTags::Detect;
     return  I18N::LangTags::Detect::detect();
 }
 
```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Sep 28 10&#8203;:22&#8203;:45 2010\, toddr@&#8203;cpanel\.net wrote&#8203;:
> 
> On Sep 27\, 2010\, at 12&#8203;:06 AM\, Father Chrysostomos via RT wrote&#8203;:
> 
> > On Thu Jul 22 10&#8203;:33&#8203;:05 2010\, toddr@&#8203;cpanel\.net wrote&#8203;:
> >> This is a bug report for perl from toddr@&#8203;cpanel\.net\,
> >> generated with the help of perlbug 1\.39 running under perl 5\.12\.1\.
> >> 
> >> 
> >> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> >> \[Please describe your issue here\]
> >> 
> >> This patch with tests resolves CPAN RT \#48808\.
> > 
> > What tests? :\-\)
> 
> My bad\. No tests\. Do you think it's needed for this change?

No\, I don’t\.

> Also I'm adding a git patch to make your life easier to commit it\.
> 

Thank you\. This has been applied as
ed897e29a200fd3c758cb7ab8c0d953347a98056\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
